Step5 : 자동차 경주 (리팩터링)

hbm0411 opened the pull request:
Step4의 리뷰를 반영하여 리팩토링 완료하였습니다.

hbm0411 committed:
refactor: step4 리뷰 반영

우승자 null 체크
코틀린 컨벤션에 맞추어 코드 수정 ( if,else문에 {} 사용 )
fun interface 사용
불필요한 테스트코드 삭제
Face Constructor 방식

hbm0411 committed:
refactor: carName을 담을수 있는 CarInfo 모델 생성 및 테스트 구현

hbm0411 committed:
refactor: RaceCar에서 maxPositionCarList 가져오도록 수정

laco-dev commented on src/main/kotlin/raicing/model/CarInfo.kt:
```diff
@@ -0,0 +1,18 @@
+
+import raicing.Const
+
+data class CarInfo(val name: String) {
```
차량에 무언가 정보가 추가 될 수는 있지만, 요구사항 구조에서는 차량 이름만 사용하기 때문에
CarName으로 의미를 가져가 보면 좋을 것 같아요~

Info 라는 이름이 만능이기는 하지만... 🤔 
그래서 무슨 역할을 하는걸까? 라는 생각을 열어보기 전까지 모르는 것 같아요

hbm0411 replied:
carName이라는 객체를 만드는것도 좋지만, 
그렇게 되면 Car에서 사용할 때 car.carName.name이 되고 이런 구조가 결국 필요한가? 싶더라구요.
결국에 지금에는 name만 객체로 만들었지만 좀 더 정보가 많이 들어가는 요구사항이 될 수도 있을거 같구요.

만약에 정말 차량 이름만 가지려고 한다면 CarName을 class가 아닌 value class 로 만드는게 나을수도 있다는 생각을 헀는데..
그것도 별로 좋지 않을거라는 생각이 들었습니다.

laco-dev replied:
네, 그 관점은 트레이드 오프 영역이라고 볼 수도 있을 것 같네요.
항상 최고의 방식으로만 도메인 설계를 할 때는 비즈니스 요구사항이 너무 복잡한 경우도 많다 보니까요.

value class에서는 프로퍼티를 하나만 가지면서 value라는 이름을 자주 사용하고는 합니다.
```kotlin
value class CarName(val value: String) { ... }
```
이 과정에서도 value 라는 이름을 출력하기 위해 접근하는게 다소 애매하다면 정말 원시값처럼 toString()을 재정의 해서 value를 아예 감추는 방법도 있어요.

```kotlin
value class CarName(private val value: String) { 
   override fun toString(): String = value
}
```
이러면 단순히 출력했을 때에도 실제 이름만 나오니까 원시값 처럼 사용할 수 있어서요 ㅎ

hbm0411 replied:
value class로 변경하여 toString 을 이용하는 방식으로 수정헀습니다 :)

laco-dev commented on src/main/kotlin/raicing/model/RaceResult.kt:
```diff
@@ -3,9 +3,8 @@ package raicing.model
-        val raceResult = raceCars.last()
-        val maxPosition = raceResult.getCars().maxOf { it.position }
-        return raceResult.getCars().filter { it.position == maxPosition }
+    fun whoIsWinner(): List<Car>? {
+        val raceLastResult = raceCars.lastOrNull()
+        return raceLastResult?.getMaxPositionCars()
```
우승 차량들에 대해서 null 과 emptyList()가 같은 의미를 가질 것으로 생각이 드는데요!
이 경우는 null 보다는 명시적으로 비어있는 리스트를 반환하는 것이 좋아요.

emptyList()는 우승자가 없구나 라고 생각할 수 있는데 null은 실패한건가? 자동차가 없나? 등의 여러가지 추측으로 퍼질 수 있어서요

laco-dev commented on src/test/kotlin/raicing/model/CarInfoTests.kt:
```diff
@@ -0,0 +1,39 @@
+                .withMessage(Const.ErrorMsg.CAR_NAME_IS_TOO_LONG)
+        }
+
+        @ParameterizedTest(name = "입력값이 `{0}` 이면 성공")
+        @ValueSource(strings = ["abc", "a", "한글"])
+        fun `입력값에 5자 이하인 이름만 있으면 성공`(carNameStr: String) {
+            val result = CarInfo(carNameStr)
+            assertThat(result.name).isEqualTo(carNameStr)
+        }
```
이 테스트는 가장 먼저 만들어 져야 하지 않을까? 라는 생각이 듭니다.
5자를 초과하지 않은 경우를 명시적으로 테스트 하기 위해 작성이 된 것 같은데
`CarInfo는 이름을 가진다` 라는 기능을 테스트 한 게아닌가? 싶네요!

hbm0411 replied:
Test의 이름과 순서를 말씀해주신거 맞죠? 수정했습니다.

laco-dev commented on src/test/kotlin/raicing/model/RaceResultTests.kt:
```diff
@@ -6,55 +6,39 @@ import org.junit.jupiter.api.Test
         assertThat(actual).isEqualTo(expected)
     }
+
+    private fun RaceCar(vararg cars: Car) = RaceCar(cars.asList())
+    private fun RaceResult(vararg raceCars: RaceCar) = RaceResult(raceCars.asList())
```
Car도 string과 position을 받는 가짜 생성자를 만들어 봐도 코드를 꽤 줄일 수 있을 것 같아요
```kotlin
Car(name: Striing = "...", position: Int = 0) = Car(CarInfo(name), position)
```

laco-dev commented on src/main/kotlin/raicing/model/RaceCar.kt:
```diff
@@ -13,19 +12,13 @@ class RaceCar(private val cars: List<Car>) {
-        }
+    companion object {
+        fun of(carInfos: List<CarInfo>): RaceCar =
+            RaceCar(List(carInfos.size) { Car(carInfos[it]) })
```
carInfos.map { ... } 을 사용해서 보다 간결하게 만들어 볼 수 있어요!

hbm0411 replied:
관성에 의해 전에 썼던 코드를 재사용했네요 ㅎㅎ 수정헀습니다.

laco-dev commented on src/test/kotlin/raicing/model/CarTests.kt:
```diff
@@ -1,66 +1,38 @@
-                assertThat(afterCar.name).isEqualTo(carName)
-                assertThat(afterCar.position).isEqualTo(car.position + 1)
-            }
-        )
+        assertThat(afterCar.position).isEqualTo(car.position + 1)
```
1을 더하는 간단한 코드라도 테스트 코드에 로직을 포함하는건 피하는걸 권장드립니다.
expected와 actual 형태로 변수로 나눠서 많이 표현하고는 해요.
```kotlin
val expected = 2
val actual = car.moveWoard(alwaysTruStrategy).position

assertThat(actual).isEqualTo(expected)
```

laco-dev reviewed (changes requested):
피드백 반영 확인했습니다 ㅎ

전체적으로 구조가 이전보다 훨씬 좋아진 것 같아요.
테스트와 몇가지 간단한 의견에 대해 의견 남겨 드렸습니다. 🙏

hbm0411 committed:
refactor: step5 리뷰 반영

test 이름 명확하게 수정
expected 값 명확하게 수정
간단한 테스트 코드를 위해 fake constructor 생성하여 사용
리스트 생성 코드 간단하게 수정

hbm0411 committed:
refactor: 우승자 없을 경우 emptyList() 반환하도록 수정

우승자 없을 경우 emptyList() 반환
postion이 0인 경우 우승자 없도록 수정

laco-dev commented on src/main/kotlin/raicing/model/RaceResult.kt:
```diff
@@ -3,8 +3,14 @@ package raicing.model
+            } else {
+                raceLastResult.getMaxPositionCars(maxPosition)
+            }
+        } ?: emptyList()
```
early return과 엘비스를 이용해 만들어 보면서 보다 가독성을 개선 할 수 있을 것 같아요.
```kotlin
val result = raceCars.lastOrNull() ?: return emptyList()
val maxPosition = result.getMaxPosition()
if (maxPosition == 0) {
   return emptyList()
}
return result.getMaxPositionCars(maxPosition)
```

hbm0411 replied:
현재 팀에서는 early return을 안좋아하는 경향이 있어 사용하지 않았던거 같네요 ㅎㅎ
수정하였습니다.

laco-dev commented on src/main/kotlin/raicing/model/RaceCar.kt:
```diff
@@ -12,8 +12,9 @@ class RaceCar(private val cars: List<Car>) {
-        val maxPosition = cars.maxOf { it.position }
+    fun getMaxPosition() = cars.maxOf { it.position }
+
+    fun getMaxPositionCars(maxPosition: Int): List<Car> {
```
이 함수는 더이상 getMaxPositionCars()가 아니게 된 것 같네요!?
```kotlin
getCarsByPosition(position: Int) { ... } 
```
형태는 어떨까요? 파라미터로 maxPosition인지 아닌지는 관심이 없어졌으니까요.

hbm0411 replied:
내용자체가 바뀌어서 메소드명도 변경되는게 맞겠네요! 수정하였습니다.

laco-dev reviewed (changes requested):
피드백 반영 확인했습니다.

마지막 미션인 만큼 조금 더 개선 해보면 좋을 것 같아서 
질문 주신 내용 포함해서 간단한 의견을 추가로 남겼습니다! 🙏

hbm0411 committed:
refactor: step5 리뷰 반영

내용 수정에 따른 메소드명 변경
early return 을 사용하여 가독성 높이기

hbm0411 committed:
refactor: CarInfo를 value class CarName으로 변경하여 적용

laco-dev reviewed (approved):
피드백 반영 확인하였습니다.

진행하면서 저도 같이 성장할 수 있는 기회가 되었던 것 같습니다.
자동차 경주 미션 고생 많으셨습니다. 🎉
